### PR TITLE
hare: unstable-2023-10-23 -> unstable-2023-11-27; harec: unstable-2023-10-22 -> unstable-2023-11-29

### PR DESCRIPTION
**Pull request by onemoresuza**
## Description of changes

- hare: migrate to pkgs/by-name
- harec: migrate to pkgs/by-name
- harec: unstable-2023-10-22 -> unstable-2023-11-29
- hare: move platform and arch check to outer let-in
- hare: refactor configurePhase and reorder attributes
- hare: patch tzdata paths
- hare: unstable-2023-10-23 -> unstable-2023-11-27
- hare: set outputs "out" and "man"

This update puts `hare` right before the move to support `OpenBSD` on it, which is still having some bugs ironed out.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by onemoresuza** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>harec</li>
    <li>himitsu</li>
  </ul>
</details>

**Comment by onemoresuza** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>hare.man</li>
    <li>harec</li>
    <li>himitsu</li>
  </ul>
</details>

File: pkgs/by-name/ha/hare/001-tzdata.patch [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ // (c) Hare authors <https://harelang.org>
+ 
+ def LOCALTIME_PATH: str = "/etc/localtime";
+-def ZONEINFO_PREFIX: str = "/usr/share/zoneinfo/";
++def ZONEINFO_PREFIX: str = "@tzdata@/share/zoneinfo/";
+ 
+ // The filepath of the system's "leap-seconds.list" file, which contains UTC/TAI
+ // leap second data.
+-export def UTC_LEAPSECS_FILE: str = "/usr/share/zoneinfo/leap-seconds.list";
++export def UTC_LEAPSECS_FILE: str = "@tzdata@/share/zoneinfo/leap-seconds.list";
```
Comment (SuperSandro2000) [inline pkgs/by-name/ha/hare/001-tzdata.patch:28]: We need to wait for https://github.com/NixOS/nixpkgs/pull/265124 for this to work https://nixpk.gs/pr-tracker.html?pr=265124

File: pkgs/by-name/ha/hare/package.nix [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+let
+  # We use harec's override of qbe until 1.2 is released, but the `qbe` argument
+  # is kept to avoid breakage.
+  qbe = harec.qbeUnstable;
+  # Until #265124 is merged (it's already on staging), this override is needed.
+  tzdata' = tzdata.overrideAttrs (old: {
+    postInstall = old.postInstall + ''
+      cp leap-seconds.list $out/share/zoneinfo
+    '';
+  });
```
Comment (SuperSandro2000) [inline pkgs/by-name/ha/hare/package.nix:22]:
```suggestion
```
lets wait until staging-next is merged https://nixpk.gs/pr-tracker.html?pr=265124

File: pkgs/by-name/ha/hare/package.nix [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  nativeBuildInputs = [
+    binutils-unwrapped
+    harec
+    makeWrapper
+    qbe
+    scdoc
+  ];
+
+  buildInputs = [
+    binutils-unwrapped
```
Comment (SuperSandro2000) [inline pkgs/by-name/ha/hare/package.nix:64]: We need this in both?

File: pkgs/by-name/ha/hare/package.nix [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+  makeFlags = [
+    "HARECACHE=.cache"
+    "PREFIX=${builtins.placeholder "out"}"
+    "PLATFORM=${platform}"
+    "ARCH=${arch}"
+  ];
+
+  # Append the distribution name to the version
+  env.LOCALVER = "nix";
```
Comment (SuperSandro2000) [inline pkgs/by-name/ha/hare/package.nix:78]:
```suggestion
  env.LOCALVER = "nixpkgs";
```

File: pkgs/by-name/ha/hare/package.nix [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+  enableParallelBuilding = true;
+
+  configurePhase = ''
+    runHook preConfigure
+
+    ln -s config.example.mk config.mk
+
+    runHook postConfigure
+  '';
```
Comment (SuperSandro2000) [inline pkgs/by-name/ha/hare/package.nix:92]:
```suggestion
  preConfigure = ''
    ln -s config.example.mk config.mk
  '';
```
does this work?

File: pkgs/by-name/ha/hare/package.nix [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    let
+      binPath = lib.makeBinPath [
+        binutils-unwrapped
+        harec
+        qbe
+      ];
+    in
+    ''
+      wrapProgram $out/bin/hare --prefix PATH : ${binPath}
+    '';
```
Comment (SuperSandro2000) [inline pkgs/by-name/ha/hare/package.nix:104]:
```suggestion
  postFixup = ''
    wrapProgram $out/bin/hare \
      --prefix PATH : ${lib.makeBinPath [ binutils-unwrapped harec qbe ]}
  '';
```
lets compact this a bit

File: pkgs/by-name/ha/hare/package.nix [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+  setupHook = ./setup-hook.sh;
+
+  meta = {
+    homepage = "https://harelang.org/";
+    description =
+      "A systems programming language designed to be simple, stable, and robust";
```
Comment (SuperSandro2000) [inline pkgs/by-name/ha/hare/package.nix:111]:
```suggestion
    description = "Systems programming language designed to be simple, stable, and robust";
```

Comment (onemoresuza) [reply pkgs/by-name/ha/hare/package.nix]: Updated the description.

File: pkgs/by-name/ha/hare/001-tzdata.patch (hunk 1 above)
Comment (onemoresuza) [reply pkgs/by-name/ha/hare/001-tzdata.patch:28]: Changed this PR to draft until https://github.com/NixOS/nixpkgs/pull/265124 is merged.

**Comment by onemoresuza** [comment]
> I don't have the know-how to work with this code. Sorry but I can't be useful here.

No worries!

**Review by ghost — COMMENTED** [review]
one packages fails to build. is this expected?

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>hareThirdParty.hare-json</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>hare.man</li>
    <li>harec</li>
    <li>himitsu</li>
  </ul>
</details>


other than that LGTM. a few nits described inline.

**Comment by onemoresuza** [comment]
> one packages fails to build. is this expected?
> 
> Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 1 package marked as broken and skipped:
> 
>     * himitsu-firefox
> 
> 
> 1 package failed to build:
> 
>     * hareThirdParty.hare-json
> 
> 
> 4 packages built:
> 
>     * hare
> 
>     * hare.man
> 
>     * harec
> 
>     * himitsu
> 
> 
> other than that LGTM. a few nits described inline.

This is to be expected: I've had the same problem when packaging [`hare-toml`](https://github.com/NixOS/nixpkgs/pull/271761/files).

I'm going to create a patch upstream and a PR to update `hare-json` to this version of `Hare`.

File: pkgs/by-name/ha/hare/package.nix [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  arch =
+    if stdenv.hostPlatform.isx86_64 then "x86_64"
+    else if stdenv.hostPlatform.isAarch64 then "aarch64"
+    else if stdenv.hostPlatform.isRiscV64 then "riscv64"
+    else builtins.throw "Unsupported architecture: ${stdenv.hostPlatform.system}";
+  platform =
+    if stdenv.hostPlatform.isLinux then "linux"
+    else if stdenv.hostPlatform.isFreeBSD then "freebsd"
+    else builtins.throw "Unsupported platform: ${stdenv.hostPlatform.system}";
+in
```
[the remark this answers is not in the document]
Comment (onemoresuza) [reply pkgs/by-name/ha/hare/package.nix]: Many thanks!, @a-n-n-a-l-e-e. Your `uname` solution makes the code much easier to read.

File: pkgs/by-name/ha/hare/package.nix [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  patches = [
+    # Replace FHS paths with nix store
+    (substituteAll {
+      src = ./001-tzdata.patch;
+      tzdata = tzdata';
+    })
+  ];
+
+  nativeBuildInputs = [
+    # binutils-unwrapped
```
[the remark this answers is not in the document]
Comment (onemoresuza) [reply pkgs/by-name/ha/hare/package.nix]: Removed it.

**Review by ghost — APPROVED** [review]

**Review by pbsds — APPROVED** [review]

**Comment by pbsds** [comment]
Any reason this is marked draft?

Result of `nixpkgs-review pr 271476` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>hareThirdParty.hare-json</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>hare.man</li>
    <li>harec</li>
    <li>himitsu</li>
  </ul>
</details>

**Comment by onemoresuza** [comment]
> Any reason this is marked draft?

Still learning the ropes of `nixpkgs`. Made this PR into a draft because it is dependent on [#265124](https://github.com/NixOS/nixpkgs/pull/265124) --- merged onto `staging`, now on `staging-next` ---, but I could just change the base branch of this PR from `master` to `staging-next`, couldn't I?

**Comment by onemoresuza** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>hareThirdParty.hare-json</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>hare.man</li>
    <li>harec</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>himitsu</li>
  </ul>
</details>

**Comment by onemoresuza** [comment]
[#265124](https://github.com/NixOS/nixpkgs/pull/265124) was merged onto master, thus, this PR is ready for review.

**Comment by onemoresuza** [comment]
Commits squashed, making this PR have 4 commits:

1. move of `harec` to `pkgs/by-name`;
2. move of `hare` to `pkgs/by-name`;
3. update of `harec`;
4. update of `hare`.

**Review by ghost — APPROVED** [review]

**Comment by onemoresuza** [comment]
@pbsds, anything else needed before the merge?

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 271476` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>hareThirdParty.hare-json</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>hare.man</li>
    <li>harec</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>himitsu</li>
  </ul>
</details>

**Pull request merged**